A58-client-side-weighted-round-robin-lb-policy.md

```diff
@@ -0,0 +1,251 @@
+  if (new_eight == 0) return;
+  if (non_empty_since_ == infy) non_empty_since_ = now();
+  last_updated_ = now();
+  weight_ = new_weight;
+}
+function OnSubchannelBecomesReady() {non_empty_since_ = infy;}
+function GetWeight() {
+   if (now() - last_updated_ >= weight_expiration_period_) {
+      non_empty_since_ = infy;
+      return 0;
+   }
+   if (now() - non_empty_since_ < blackout_period_) return 0;
+   return weight_;
+}
+
+```
+
+### Per-Call Load Reporting
+
+Necessary support for per-call load report handling in clients depends on [gRFC A51]. This section outlines how this features works with `weighted_round_robin`. xDS `cluster_impl` policy can serve as an example.
+
+`weighted_round_robin` defaults to per-call load reporting. The client does nothing when the per-call load report is expected but missing in the trailing metadata of the response. While `enable_oob_load_report` is set, per-call load reports are never parsed on the client side. The client does nothing when the per-call load report is present in the response when not expected.
+
+#### C/C++
+Processing per-call load reports is supported with `SubchannelCallTrackerInterface`. `weighted_round_robin` implements this interface and attaches it to a subchannel `PickResult`. Later Finish() is called back with the load metrics and updates the weight.
+
+#### Java
+`OrcaPerRequestUtil.OrcaReportingTracerFactory` is available in `io.grpc.xds.orca` to process per-call backend load reports in `ClusterImplLoadBalancer`. This needs to be moved to a new package and shared with `weighted_round_robin`. Then the picker attaches this tracer to the subchannel selection and the listener takes the reference to the subchannel to which it updates the weight to in `onLoadReport()`.
+
+#### Go
+Per-call load report handling is fully implemented. Picker selection result `PickResult` has an optional callback `Done()` that is passed with the backend load metrics parsed from the response if available. With `weighted_round_robin`, the picker only sets Done() in `PickResult` when per-call load reporting is enabled, and the callback updates the weight.
+
+### OOB Load Reporting
+
+OOB load report handling in clients depends on [gRFC A51]. This section outlines how this feature works with `weighted_round_robin`.
+
+In all languages, `report_interval` in `OrcaLoadReportRequest` is determined by the service config parameter `oob_reporting_period`. When it changes, a new stream request is sent. In Java/Go, the implementation needs to add support to enable and disable the OOB listener when `enable_oob_load_report` is updated. In C++ this happens automatically as a new subchannel list is created.
+
+`weighted_round_robin` initiates OOB load reporting when the LB policy parameter `enable_oob_load_report` changes to true. When the backend returns `UNIMPLEMENTED` to the OOB load report request it proceeds with a log message as if it is not enabled and does not retry.
```

Comment:
The second sentence here is not needed, because that's already covered in gRFC A51.

Suggest just removing this paragraph in favor of the alternative wording I suggested above.

Reply:
Done.